p2p/server.py

```diff
@@ -0,0 +1,156 @@
+import asyncio
+import os
+import logging
+import secrets
+import sys
+
+from typing import Tuple, cast
+from eth_keys import (
+    datatypes,
+    keys,
+)
+from eth_utils import (
+    decode_hex,
+    keccak,
+)
+
+from p2p.auth import (
+    HandshakeResponder,
+    decode_auth_plain,
+    decode_auth_eip8,
+)
+from p2p.cancel_token import CancelToken
+from p2p.constants import (
+    ENCRYPTED_AUTH_MSG_LEN,
+    HASH_LEN,
+)
+from p2p.ecies import ecdh_agree
+from p2p.exceptions import OperationCancelled
+from p2p.kademlia import (
+    Address,
+    Node,
+)
+from p2p.peer import PeerPool, BasePeer, PeerPoolSubscriber, ETHPeer, LESPeer
+from p2p.utils import sxor
+
+
+class Server(PeerPoolSubscriber):
+    """Server listening for incoming connections"""
+    logger = logging.getLogger("p2p.server.Server")
+    
+    def __init__(self, privkey: datatypes.PrivateKey, server_addr: Tuple[str, str], peer_pool: PeerPool) -> None:
+        self.cancel_token = CancelToken('Server')
+        self.incoming_connections = []
+        self.peer_pool = peer_pool
+        self.peer_pool.subscribe(self)
+        self._running_peers = set()
+        self.privkey = privkey
+        self._server_address = server_addr
+
+    def register_peer(self, peer: BasePeer) -> None:
+        asyncio.ensure_future(self.handle_peer(cast(ETHPeer, peer)))
+
+    async def handle_peer(self, peer: ETHPeer) -> None:
+        """Handle the lifecycle of the given peer."""
+        self._running_peers.add(peer)
+        try:
+            await self._handle_peer(peer)
+        finally:
+            self._running_peers.remove(peer)
+
+    async def _handle_peer(self, peer: ETHPeer) -> None:
+        while True:
+            try:
+                cmd, msg = await peer.read_sub_proto_msg(self.cancel_token)
+            except OperationCancelled:
+                break
+
+    async def stop(self) -> None:
+        self.logger.info("Closing server...")
+        self.cancel_token.trigger()
+        self.peer_pool.unsubscribe(self)
```

Comment:
Is it necessary to stop the peer_pool here as well? Something like `peer_pool.stop()`?

Reply:
I don't believe so, I think the assumption is that the peer pool runs at a higher level than this process.

Reply:
Actually, my current thinking is that `Server` represents the node's main process and manages all services needed to run a full node (e.g. finding/connecting-to remote peers, accepting incoming connections, syncing the chain, etc), so yes it would be responsible for starting/stopping `PeerPool`